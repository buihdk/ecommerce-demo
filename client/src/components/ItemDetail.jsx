import React, { memo, lazy } from 'react';
import PropTypes from 'prop-types';
import {
  makeStyles,
  Paper,
  Box,
  CardMedia,
  Chip,
  Avatar,
  Typography,
} from '@material-ui/core';
import { FavoriteBorder, ChatBubbleOutline, Flag } from '@material-ui/icons';

const useStyles = makeStyles({
  media: {
    height: 400,
  },
  icon: {
    fontSize: 16,
  },
  chipWrapper: {
    marginTop: 16,
    marginBottom: 8,
  },
  count: {
    color: 'darkgrey',
    fontWeight: 600,
    lineHeight: '32px',
    display: 'inline-block',
    marginLeft: 4,
  },
  body: {
    padding: '0px 20px 8px',
  },
  footer: {
    color: 'white',
    backgroundColor: '#35231F',
    display: 'flex',
    alignItems: 'baseline',
    padding: 10,
    position: 'relative',
  },
  footerCallToAction: {
    backgroundColor: '#e43',
    position: 'absolute',
    top: 10,
    right: 10,
    lineHeight: '39px',
    padding: '0px 10px',
  },
});

const ItemDetail = ({ item, categoryName }) => {
  const styles = useStyles();

  if (!item) {
    const NotFoundPage = lazy(() => import('utils/NotFoundPage'));
    return <NotFoundPage errorCode={400} />;
  }

  return (
    <Paper>
      <Typography variant="h6" align="center">
        {item.name}
      </Typography>
      <CardMedia
        className={styles.media}
        image={item.image ? item.image : 'https://picsum.photos/id/0/400/400'}
      />
      <Typography className={styles.body}>{`Name: ${item.name}`}</Typography>
      <Box
        className={styles.body}
        display="flex"
        justifyContent="space-between"
      >
        <Box className={styles.chipWrapper}>
          <Chip
            label="いいね!"
            icon={<FavoriteBorder className={styles.icon} />}
          />
          <Box className={styles.count}>{item.like_count}</Box>
        </Box>
        <Box className={styles.chipWrapper}>
          <Chip
            label="コメント"
            icon={<ChatBubbleOutline className={styles.icon} />}
          />
          <Box className={styles.count}>{item.comment_count}</Box>
        </Box>
        <Box
          className={styles.chipWrapper}
          display={{ xs: 'none', sm: 'block' }}
        >
          <Avatar>
            <Flag />
          </Avatar>
        </Box>
      </Box>
      <Typography className={styles.body}>
        {`Categories: ${categoryName}`}
      </Typography>
      <Typography className={styles.body}>
        {`Description: ${item.description}`}
      </Typography>
      <Box className={styles.footer}>
        <Typography component="span" variant="h4">
          {`¥ ${item.price}`}
        </Typography>
        <Typography
          component="span"
          variant="subtitle2"
          style={{ marginLeft: 8 }}
        >
          {item.shipping_fee}
        </Typography>
        <Box
          className={styles.footerCallToAction}
          display={{ xs: 'none', sm: 'block' }}
        >
          カートに追加
        </Box>
      </Box>
    </Paper>
  );
};

ItemDetail.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.string,
    image: PropTypes.string,
    name: PropTypes.string,
    description: PropTypes.string,
    shipping_fee: PropTypes.string,
    price: PropTypes.number,
    like_count: PropTypes.number,
    comment_count: PropTypes.number,
    is_sold_out: PropTypes.bool,
  }),
  categoryName: PropTypes.string,
};

ItemDetail.defaultProps = {
  item: {
    id: '',
    image: '',
    name: '',
    description: '',
    shipping_fee: '',
    price: 0,
    like_count: 0,
    comment_count: 0,
    is_sold_out: false,
  },
  categoryName: '',
};

export default memo(ItemDetail);
