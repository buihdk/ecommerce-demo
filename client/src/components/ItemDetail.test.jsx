import React, { Suspense } from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { renderHook } from '@testing-library/react-hooks';

import ItemDetail, { useHistoryPush } from './ItemDetail';

const mockHistory = { push: jest.fn() };
jest.mock('react-router-dom', () => ({ useHistory: () => mockHistory }));

const item = {
  id: '1',
  image: '1',
  name: 'item 1',
  description: 'description 1',
  shipping_fee: 'free',
  price: 10,
  like_count: 2,
  comment_count: 3,
  is_sold_out: false,
};

const newItem = {
  ...item,
  image: '',
};

describe(`useHistoryPush`, () => {
  test(`works correctly`, () => {
    const {
      result: { current: handleClick },
    } = renderHook(() => useHistoryPush(mockHistory));
    handleClick();

    expect(mockHistory.push).toBeCalledWith('/');
  });
});

describe('ItemDetail', () => {
  test('renders null item without crashing', () => {
    const wrapper = mount(
      <Suspense fallback={<div>loading...</div>}>
        <ItemDetail item={null} />
      </Suspense>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('renders valid item without crashing', () => {
    const wrapper = shallow(<ItemDetail item={item} categoryName="men" />);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test('renders item with no image without crashing', () => {
    const wrapper = shallow(<ItemDetail item={newItem} categoryName="baby" />);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
