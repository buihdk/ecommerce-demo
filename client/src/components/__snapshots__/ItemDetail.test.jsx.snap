// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetail renders item with no image without crashing 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-container-1"
  >
    <WithStyles(ForwardRef(Typography))
      align="center"
      variant="h6"
    >
      item 1
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(CardMedia))
      className="makeStyles-media-2"
      image="https://picsum.photos/id/0/400/400"
    />
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-body-6"
    >
      Name: item 1
    </WithStyles(ForwardRef(Typography))>
    <Styled(MuiBox)
      className="makeStyles-body-6"
      display="flex"
      justifyContent="space-between"
    >
      <Styled(MuiBox)
        className="makeStyles-chipWrapper-4"
      >
        <WithStyles(ForwardRef(Chip))
          icon={
            <UNDEFINED
              className="makeStyles-icon-3"
            />
          }
          label="いいね!"
        />
        <Styled(MuiBox)
          className="makeStyles-count-5"
        >
          2
        </Styled(MuiBox)>
      </Styled(MuiBox)>
      <Styled(MuiBox)
        className="makeStyles-chipWrapper-4"
      >
        <WithStyles(ForwardRef(Chip))
          icon={
            <UNDEFINED
              className="makeStyles-icon-3"
            />
          }
          label="コメント"
        />
        <Styled(MuiBox)
          className="makeStyles-count-5"
        >
          3
        </Styled(MuiBox)>
      </Styled(MuiBox)>
      <Styled(MuiBox)
        className="makeStyles-chipWrapper-4"
        display={
          Object {
            "sm": "block",
            "xs": "none",
          }
        }
      >
        <WithStyles(ForwardRef(Avatar))>
          <FlagIcon />
        </WithStyles(ForwardRef(Avatar))>
      </Styled(MuiBox)>
    </Styled(MuiBox)>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-body-6"
    >
      Categories: baby
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-body-6"
    >
      Description: description 1
    </WithStyles(ForwardRef(Typography))>
    <Styled(MuiBox)
      className="makeStyles-footer-7"
    >
      <WithStyles(ForwardRef(Typography))
        component="span"
        variant="h4"
      >
        ¥ 10
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        component="span"
        style={
          Object {
            "marginLeft": 8,
          }
        }
        variant="subtitle2"
      >
        free
      </WithStyles(ForwardRef(Typography))>
      <Styled(MuiBox)
        className="makeStyles-footerCallToAction-8"
        display={
          Object {
            "sm": "block",
            "xs": "none",
          }
        }
      >
        カートに追加
      </Styled(MuiBox)>
    </Styled(MuiBox)>
  </WithStyles(ForwardRef(Paper))>
  <WithStyles(ForwardRef(Fab))
    className="makeStyles-backBtn-9"
    onClick={[Function]}
    variant="extended"
  >
    <ArrowBackIosIcon />
    Back
  </WithStyles(ForwardRef(Fab))>
</Fragment>
`;

exports[`ItemDetail renders null item without crashing 1`] = `ReactWrapper {}`;

exports[`ItemDetail renders valid item without crashing 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Paper))
    className="makeStyles-container-1"
  >
    <WithStyles(ForwardRef(Typography))
      align="center"
      variant="h6"
    >
      item 1
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(CardMedia))
      className="makeStyles-media-2"
      image="1"
    />
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-body-6"
    >
      Name: item 1
    </WithStyles(ForwardRef(Typography))>
    <Styled(MuiBox)
      className="makeStyles-body-6"
      display="flex"
      justifyContent="space-between"
    >
      <Styled(MuiBox)
        className="makeStyles-chipWrapper-4"
      >
        <WithStyles(ForwardRef(Chip))
          icon={
            <UNDEFINED
              className="makeStyles-icon-3"
            />
          }
          label="いいね!"
        />
        <Styled(MuiBox)
          className="makeStyles-count-5"
        >
          2
        </Styled(MuiBox)>
      </Styled(MuiBox)>
      <Styled(MuiBox)
        className="makeStyles-chipWrapper-4"
      >
        <WithStyles(ForwardRef(Chip))
          icon={
            <UNDEFINED
              className="makeStyles-icon-3"
            />
          }
          label="コメント"
        />
        <Styled(MuiBox)
          className="makeStyles-count-5"
        >
          3
        </Styled(MuiBox)>
      </Styled(MuiBox)>
      <Styled(MuiBox)
        className="makeStyles-chipWrapper-4"
        display={
          Object {
            "sm": "block",
            "xs": "none",
          }
        }
      >
        <WithStyles(ForwardRef(Avatar))>
          <FlagIcon />
        </WithStyles(ForwardRef(Avatar))>
      </Styled(MuiBox)>
    </Styled(MuiBox)>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-body-6"
    >
      Categories: men
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-body-6"
    >
      Description: description 1
    </WithStyles(ForwardRef(Typography))>
    <Styled(MuiBox)
      className="makeStyles-footer-7"
    >
      <WithStyles(ForwardRef(Typography))
        component="span"
        variant="h4"
      >
        ¥ 10
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        component="span"
        style={
          Object {
            "marginLeft": 8,
          }
        }
        variant="subtitle2"
      >
        free
      </WithStyles(ForwardRef(Typography))>
      <Styled(MuiBox)
        className="makeStyles-footerCallToAction-8"
        display={
          Object {
            "sm": "block",
            "xs": "none",
          }
        }
      >
        カートに追加
      </Styled(MuiBox)>
    </Styled(MuiBox)>
  </WithStyles(ForwardRef(Paper))>
  <WithStyles(ForwardRef(Fab))
    className="makeStyles-backBtn-9"
    onClick={[Function]}
    variant="extended"
  >
    <ArrowBackIosIcon />
    Back
  </WithStyles(ForwardRef(Fab))>
</Fragment>
`;
